// ix/asynciterable/pipe/orderby.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const orderby_1 = require("../orderby");
const sorter_1 = require("../../internal/sorter");
function orderBy(keySelector, comparer = sorter_1.sorter) {
    return function orderByOperatorFunction(source) {
        return new orderby_1.OrderedAsyncIterableX(source, keySelector, comparer, false);
    };
}
exports.orderBy = orderBy;
function orderByDescending(keySelector, comparer = sorter_1.sorter) {
    return function orderByDescendingOperatorFunction(source) {
        return new orderby_1.OrderedAsyncIterableX(source, keySelector, comparer, true);
    };
}
exports.orderByDescending = orderByDescending;
function thenBy(keySelector, comparer = sorter_1.sorter) {
    return function thenByOperatorFunction(source) {
        const orderSource = source;
        return new orderby_1.OrderedAsyncIterableX(orderSource._source, keySelector, comparer, false, orderSource);
    };
}
exports.thenBy = thenBy;
function thenByDescending(keySelector, comparer = sorter_1.sorter) {
    return function thenByDescendingOperatorFunction(source) {
        const orderSource = source;
        return new orderby_1.OrderedAsyncIterableX(orderSource._source, keySelector, comparer, true, orderSource);
    };
}
exports.thenByDescending = thenByDescending;
