// ix/internal/isiterable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** @ignore */
const isNumber = (x) => typeof x === 'number';
/** @ignore */
const isBoolean = (x) => typeof x === 'boolean';
/** @ignore */
exports.isFunction = (x) => typeof x === 'function';
/** @ignore */
exports.isObject = (x) => x != null && Object(x) === x;
/** @ignore */
exports.isPromise = (x) => {
    return exports.isObject(x) && exports.isFunction(x.then);
};
/** @ignore */
function isArrayLike(x) {
    return exports.isObject(x) && isNumber(x['length']);
}
exports.isArrayLike = isArrayLike;
/** @ignore */
function isIterable(x) {
    return exports.isObject(x) && exports.isFunction(x[Symbol.iterator]);
}
exports.isIterable = isIterable;
/** @ignore */
function isIterator(x) {
    return exports.isObject(x) && !exports.isFunction(x[Symbol.iterator]) && exports.isFunction(x['next']);
}
exports.isIterator = isIterator;
/** @ignore */
function isAsyncIterable(x) {
    return exports.isObject(x) && exports.isFunction(x[Symbol.asyncIterator]);
}
exports.isAsyncIterable = isAsyncIterable;
/** @ignore */
exports.isReadableNodeStream = (x) => {
    return (exports.isObject(x) &&
        exports.isFunction(x['pipe']) &&
        exports.isFunction(x['_read']) &&
        isBoolean(x['readable']) &&
        exports.isObject(x['_readableState']));
};
/** @ignore */
exports.isWritableNodeStream = (x) => {
    return (exports.isObject(x) &&
        exports.isFunction(x['end']) &&
        exports.isFunction(x['_write']) &&
        isBoolean(x['writable']) &&
        exports.isObject(x['_writableState']));
};
/** @ignore */
exports.isReadableDOMStream = (x) => {
    return exports.isObject(x) && exports.isFunction(x['cancel']) && exports.isFunction(x['getReader']);
};
/** @ignore */
exports.isWritableDOMStream = (x) => {
    return exports.isObject(x) && exports.isFunction(x['abort']) && exports.isFunction(x['getWriter']);
};
/** @ignore */
exports.isFetchResponse = (x) => {
    return exports.isObject(x) && exports.isReadableDOMStream(x['body']);
};
