// ix/Ix.dom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const iterableX = require("./iterable/index");
exports.iterable = iterableX;
const iterableXPipe = require("./iterable/pipe/index");
exports.iterablePipe = iterableXPipe;
const asynciterableX = require("./asynciterable/index");
exports.asynciterable = asynciterableX;
const asynciterableXPipe = require("./asynciterable/pipe/index");
exports.asynciterablePipe = asynciterableXPipe;
// Manually re-export because closure-compiler doesn't support `export * from X` syntax yet
var Ix_1 = require("./Ix");
exports.default = Ix_1.default;
var Ix_2 = require("./Ix");
exports.OrderedIterable = Ix_2.OrderedIterable;
var Ix_3 = require("./Ix");
exports.OrderedIterableBase = Ix_3.OrderedIterableBase;
var Ix_4 = require("./Ix");
exports.OrderedAsyncIterable = Ix_4.OrderedAsyncIterable;
var Ix_5 = require("./Ix");
exports.OrderedAsyncIterableBase = Ix_5.OrderedAsyncIterableBase;
var Ix_6 = require("./Ix");
exports.AsyncSink = Ix_6.AsyncSink;
exports.Iterable = Ix_6.Iterable;
exports.AsyncIterable = Ix_6.AsyncIterable;
require("./add/asynciterable/fromdomstream");
require("./add/iterable-operators/todomstream");
require("./add/asynciterable-operators/todomstream");
var fromdomstream_1 = require("./asynciterable/fromdomstream");
exports.fromDOMStream = fromdomstream_1.fromDOMStream;
exports.AsyncIterableReadableStream = fromdomstream_1.AsyncIterableReadableStream;
exports.AsyncIterableReadableByteStream = fromdomstream_1.AsyncIterableReadableByteStream;
