// ts-prime/dist/commonjs/guards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore
function isString(data) {
    return typeof data === 'string';
}
exports.isString = isString;
// @ts-ignore
function isNumber(data) {
    return typeof data === 'number' && !isNaN(data);
}
exports.isNumber = isNumber;
function isDefined(data) {
    return typeof data !== 'undefined' && data !== null;
}
exports.isDefined = isDefined;
// @ts-ignore
function isBoolean(data) {
    return typeof data === 'boolean';
}
exports.isBoolean = isBoolean;
// @ts-ignore
function isPromise(data) {
    return data instanceof Promise;
}
exports.isPromise = isPromise;
// @ts-ignore
function isArray(data) {
    return Array.isArray(data);
}
exports.isArray = isArray;
// @ts-ignore
function isObject(data) {
    return !!data && !Array.isArray(data) && typeof data === 'object';
}
exports.isObject = isObject;
// @ts-ignore
function isFunction(data) {
    return typeof data === 'function';
}
exports.isFunction = isFunction;
function isNil(data) {
    return data == null;
}
exports.isNil = isNil;
// @ts-ignore
function isError(data) {
    return data instanceof Error;
}
exports.isError = isError;
// @ts-ignore
function isDate(data) {
    return data instanceof Date;
}
exports.isDate = isDate;
function isTruthy(value) {
    return !!value;
}
exports.isTruthy = isTruthy;
function isNot(predicate) {
    return function (data) {
        return !predicate(data);
    };
}
exports.isNot = isNot;
