// ix/iterable/iterablex.mjs
import { bindCallback } from '../internal/bindcallback';
import { identity } from '../internal/identity';
import { toLength } from '../internal/tolength';
import { isArrayLike, isIterable, isIterator, isReadableNodeStream, isWritableNodeStream } from '../internal/isiterable';
/**
 * This clas serves as the base for all operations which support [Symbol.iterator].
 */
export class IterableX {
    forEach(projection, thisArg) {
        const fn = bindCallback(projection, thisArg, 2);
        let i = 0;
        for (let item of this) {
            fn(item, i++);
        }
    }
    pipe(...args) {
        let i = -1;
        let n = args.length;
        let acc = this;
        let as = IterableX.as;
        while (++i < n) {
            acc = as(args[i](acc));
        }
        return acc;
    }
    tee() {
        return this._getDOMStream().tee();
    }
    pipeTo(writable, options) {
        return this._getDOMStream().pipeTo(writable, options);
    }
    pipeThrough(duplex, options) {
        return this._getDOMStream().pipeThrough(duplex, options);
    }
    _getDOMStream() {
        return this._DOMStream || (this._DOMStream = this.publish().toDOMStream());
    }
    /** @nocollapse */
    static as(source) {
        /* tslint:disable */
        if (source instanceof IterableX) {
            return source;
        }
        if (typeof source === 'string') {
            return new OfIterable([source]);
        }
        if (isIterable(source)) {
            return new FromIterable(source, identity);
        }
        if (isArrayLike(source)) {
            return new FromIterable(source, identity);
        }
        return new OfIterable([source]);
        /* tslint:enable */
    }
    /** @nocollapse */
    static from(source, selector = identity, thisArg) {
        const fn = bindCallback(selector, thisArg, 2);
        /* tslint:disable */
        if (isIterable(source)) {
            return new FromIterable(source, fn);
        }
        if (isArrayLike(source)) {
            return new FromIterable(source, fn);
        }
        if (isIterator(source)) {
            return new FromIterable({ [Symbol.iterator]: () => source }, fn);
        }
        throw new TypeError('Input type not supported');
        /* tslint:enable */
    }
    /** @nocollapse */
    static of(...args) {
        //tslint:disable-next-line
        return new OfIterable(args);
    }
}
class FromIterable extends IterableX {
    constructor(source, fn) {
        super();
        this._source = source;
        this._fn = fn;
    }
    *[Symbol.iterator]() {
        const iterable = isIterable(this._source);
        let i = 0;
        if (iterable) {
            for (let item of this._source) {
                yield this._fn(item, i++);
            }
        }
        else {
            let length = toLength(this._source.length);
            while (i < length) {
                let val = this._source[i];
                yield this._fn(val, i++);
            }
        }
    }
}
class OfIterable extends IterableX {
    constructor(args) {
        super();
        this._args = args;
    }
    *[Symbol.iterator]() {
        yield* this._args;
    }
}
try {
    (isBrowser => {
        if (isBrowser) {
            return;
        }
        const as = IterableX.as;
        IterableX.prototype.pipe = nodePipe;
        const readableOpts = (x, opts = x._writableState || { objectMode: true }) => opts;
        function nodePipe(...args) {
            let i = -1;
            let end;
            let n = args.length;
            let prev = this;
            let next;
            while (++i < n) {
                next = args[i];
                if (typeof next === 'function') {
                    prev = as(next(prev));
                }
                else if (isWritableNodeStream(next)) {
                    ({ end = true } = args[i + 1] || {});
                    // prettier-ignore
                    return isReadableNodeStream(prev) ? prev.pipe(next, { end }) :
                        prev.toNodeStream(readableOpts(next)).pipe(next, { end });
                }
            }
            return prev;
        }
    })(typeof window === 'object' && typeof document === 'object' && document.nodeType === 9);
}
catch (e) {
    /* */
}
