// ix/asynciterable/toobservable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const toobserver_1 = require("../internal/toobserver");
const observer_1 = require("../observer");
class BooleanSubscription {
    constructor() {
        this.isUnsubscribed = false;
    }
    unsubscribe() {
        this.isUnsubscribed = true;
    }
}
class AsyncIterableObservable {
    constructor(source) {
        this._source = source;
    }
    [observer_1.observable]() {
        return this;
    }
    subscribe(next, error, complete) {
        const observer = toobserver_1.toObserver(next, error, complete);
        const subscription = new BooleanSubscription();
        const it = this._source[Symbol.asyncIterator]();
        const f = () => {
            it.next()
                .then(({ value, done }) => {
                if (!subscription.isUnsubscribed) {
                    if (done) {
                        observer.complete();
                    }
                    else {
                        observer.next(value);
                        f();
                    }
                }
            })
                .catch(err => {
                if (!subscription.isUnsubscribed) {
                    observer.error(err);
                }
            });
        };
        f();
        return subscription;
    }
}
function toObservable(source) {
    return new AsyncIterableObservable(source);
}
exports.toObservable = toObservable;
